### ISSUE TITLE
[Bug] Nodes are too relaxed on block's timestamp

### ISSUE THREAD

**Issue by HarukaMa** [AleoHQ/snarkOS#1379]
## Foreword

It looks like this issue is being widely circulated, so I'd like to add a few explanation here.

1. Some people are cheating

It depends on how you see it. The current consensus code allows future timestamps as long as it's still within two hours in the future, and I just think this makes no sense considering how mining works.

2. This is the "answer" why the testnet2 is in this state

Actually what's happening on testnet2 is just a 51% - some or one person have more mining power than everyone else. *Fixing this issue won't make them stop.* They can still mine blocks with high weights and only release the blocks after the timestamp in the block is reasonable. They will still monopolize the testnet until pools or better miners appear.

## 🐛 Bug Report

As this is still in the testnet, I don't consider it a "security issue" so I'm publicly reporting this.

As we can see, there are miners generating blocks with a constant `difficulty_target`: `922115695706806528`. If you check the block though, you will see:

```
$ curl 5.61.53.57:3032 -d '{"jsonrpc": "2.0", "id":"", "method": "latestblockheader", "params": [] }' | jq .result.metadata
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1726  100  1653  100    73   5251    231 --:--:-- --:--:-- --:--:--  5479
{
  "cumulative_weight": 1294868,
  "difficulty_target": 922115695706806528,
  "height": 16533,
  "timestamp": 1638686585
}
```

That timestamp is **in the future** as I write this report. It looks like those miners are forcing the `block_timestamp` parameter to force the next block to have a specific difficulty target.

Although time syncing is difficult, the nodes should not accept blocks which has invalid timestamps.

EDIT: after checking the code, it seems the snarkVM is checking if the block is within two hours in the future. However, it leaves some issues for this kind of manipulations and two hours is WAY too long for a machine with proper time. Besides, when mining the block the timestamp is determined first, so it's generally not possible for the timestamp to be in the future.

## Steps to Reproduce

Check some nodes with a high weight. Use aleochain.io is helpful.

## Expected Behavior

Blocks with timestamp in the future should be rejected.

## Your Environment

N/A

**Comment by Clarxxon** [comment]
yes, this is a very strange bug. It looks like some miners have monopolized testnet

**Comment by HarukaMa** [comment]
candidate fix: #1383

**Comment by raychu86** [comment]
Closing with #1383. Currently reduced to 2 minutes for the maximum future block timestamp. This value can be constrained further in the future once the network has matured.
